build_matrix_A.py

```diff
@@ -0,0 +1,91 @@
+    various layers of graphite)
+    - mesh: the physical location of each mesh point
+    - materials: the materials in each region and how many nodes are assigned to that material
+    outputs. List of tuples of format (material, number of nodes)
+    - time: An array of timesteps taken during transients. If steady state, supply 0.
+    --------
+    - An array that can be plugged into a subsequent function to solve the
+      A.x=b formula
+    """
+    k = np.zeros(len(mesh))
+    rho = np.zeros(len(mesh))
+    c = np.zeros(len(mesh))
+    DR = np.zeros(len(mesh))
+    if len(time) > 1:
+    	Dt = time[1] - time[0]
+    else:
+    	Dt = 1
+    for m in range(len(mesh)):
+        k[m] = material_property_library[mesh[m][0]]['k']
+        rho[m] = material_property_library[mesh[m][0]]['rho']
+        c[m] = material_property_library[mesh[m][0]]['c']
+        DR[m] = mesh[m][1][1]-mesh[m][1][0]
+
+
+    total_nodes = len(mesh[0][1])
+    for i in range(1, len(mesh)):
+        total_nodes += len(mesh[i][1])-1
+        
+    A = np.zeros((total_nodes, total_nodes))
+    volume = np.zeros(total_nodes)
+    
+    "Begin by setting boundary nodes"
+    "Have to treat nodes 0 and 1 differently to avoid divide by zero"
+    volume[0] = 4/3*np.pi*(mesh[0][1][0]+DR[0]/2)**3
+    volume[1] = 4/3*np.pi*(mesh[0][1][1]+DR[0]/2)**3-(mesh[0][1][1]-DR[0]/2)**3
+    
+    A[0, 0] = k[0]*Dt*4*np.pi/(rho[0]*c[0])/(DR[0]*volume[0])*(DR[0]/2)**2
+    A[0, 1] = -k[0]*Dt*4*np.pi/(rho[0]*c[0])/(DR[0]*volume[0])*(DR[0]/2)**2
+	
+
+    A[1, 1] = k[0]*Dt*4*np.pi/(rho[0]*c[0])/volume[1]*(1/(1/mesh[0][1][1]-1/mesh[0][1][2]) +
+                                    1/DR[0]*(mesh[0][1][0]+DR[0]/2)**2)
+    A[1, 0] = -k[0]*Dt*4*np.pi/(rho[0]*c[0])/(DR[0]*volume[1])*(mesh[0][1][0]+DR[0]/2)**2
+    A[1, 2] = -k[0]*Dt*4*np.pi/(rho[0]*c[0])/volume[1]*(1/(1/mesh[0][1][1]-1/mesh[0][1][2]))
+	
+
+	
+    for i in range(2, len(mesh[0][1])-1): #Length set to fueled node length
+        volume[i] = 4/3*np.pi*((mesh[0][1][i]+DR[0]/2)**3-(mesh[0][1][i]-DR[0]/2)**3)
+        A[i, i] = k[0]*Dt*4*np.pi/(rho[0]*c[0])/volume[i]*(1/(1/mesh[0][1][i]-1/mesh[0][1][i+1]) +
+                                        1/(1/mesh[0][1][i-1]-1/mesh[0][1][i]))
+        A[i, i-1] = -k[0]*Dt*4*np.pi/(rho[0]*c[0])/volume[i]*(1/(1/mesh[0][1][i-1]-1/mesh[0][1][i]))
+        A[i, i+1] = -k[0]*Dt*4*np.pi/(rho[0]*c[0])/volume[i]*(1/(1/mesh[0][1][i]-1/mesh[0][1][i+1]))
+		
+    #Now need to handle the boundary nodes
+    #i is currently the value of the node at the internal boundary
+    for mat in range(1, len(mesh)):
+        prev_mesh_length = len(mesh[mat-1][1])
+        volume[i+1] = 4/3*np.pi*((mesh[mat-1][1][prev_mesh_length-1]+DR[mat]/2)**3-(
+                                                 mesh[mat-1][1][prev_mesh_length-1]-DR[mat-1]/2)**3)
+        A[i+1,i+2] = -k[mat]*Dt*4*np.pi/(rho[mat-1]*c[mat-1])/volume[i+1]*(1/(1/mesh[mat][1][0]-1/mesh[mat][1][1]))
```

Comment:
should this be `k[mat-1]` to match the indexing of `rho` and `c`?

Would it be easier to recompute `k/(rho * c)` for each material?

Reply:
This is somewhat of a sticking point. It should be k, rho, and c of mat, not mat-1. But in order for things to properly cancel out with the b matrix, I had to cheat and change rho and c to mat-1. If I don't do this the program produces nonsensical answers. Since the emphasis of the project wasn't on building an accurate temperature model and we were running low on time I fudged it a bit to get it to a point where it produces somewhat valid results.

Reply:
Thanks for the clarification on this Paul, I think the weighted implementation is (more) accurate now.